_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+
+The other reason `react-responsive-media` renders _all_ breakpoints has to do with the hydration of a
+server-rendered React app. When using SSR in a React app, you need to call
+[`ReactDOM.hydrate()`](https://reactjs.org/docs/react-dom.html#hydrate) to connect your server-rendered app to the
+client's browser. As the docs describe,
+
+> React expects that the rendered content is identical between the server and the client.
+
+If there are differences between your server-rendered component tree and the initial client-rendered component
+tree, a couple things will happen.
+
+1. You'll get a friendly error in your console telling you there were differences (in development mode), but
+   also...
+2. weird, unpredictable things will happen to your app.
+
+Here's what our app looked like in one case where our client render didn't match our server render:
+
+(TODO: image of spinning body)
+
+Yuck! The docs explain that `ReactDOM.hydrate` doesn't put a lot of effort into patching up differences between the
+server and client renders, and that's how you end up with weird things like above.
+
+Because hydration is so picky about the component trees matching between server and client, we render _all_
+breakpoints from the server with `react-responsive-media`. This way, we know you're going to get a branch that
+matches between the client and server. Then, upon hydration, we prune the branches that don't match the current
+breakpoint.
+
+## Tool 3: [`@artsy/detect-responsive-traits`](https://github.com/artsy/detect-responsive-traits)
+
+I mentioned above that it's difficult to accurately detect devices by user agent to identify which breakpoint to
+render. We didn't want this to be our primary strategy for combining SSR with responsive design.
+
+But with `react-responsive-media` as our primary approach, we felt that we could make some further optimizations
+with user agent detection. In the event that we don't know your device by its user agent, we'll still give you both
+the desktop and mobile breakpoints from SSR. But if we are certain you are on a mobile device, we can omit the
+desktop breakpoint from the server render.
+
+We really wanted to not maintain our own list of user agents. Alas, we found that none of the existing user agent
+detection libraries surfaced all the information we needed. We needed to know the minimum width for a browser on a
+given device, and if it was resizable, and to what dimensions it was resizable. If any existing libraries _did_
+have this data, they didn't provide it to us easily.
+
+So we did some experimentation, given the browsers & devices we knew we needed to support. And yeah...we
+(reluctantly) created our own user-agent detection library,
+[@artsy/detect-responsive-traits](https://github.com/artsy/detect-responsive-traits). This data is very specific to
+the browsers & devices we support on [artsy.net](artsy.net), but it's the data we need to optimize our responsive
```

Comment:
```suggestion
the browsers and devices we support on [artsy.net](artsy.net), but it's the data we need to optimize our responsive
```

Reply:
@pepopowitz I don't necessarily think you have to merge these if you just consider this part of your style. Just adding these as an easy one-click thing if you decide to update.

Reply:
No, I'm glad you're recommending these. It is my "style" in the sense that I often use `&` in notes, but wouldn't want to ship a blog article that used `&`.